src/ddmd/dcache.d

```diff
@@ -0,0 +1,376 @@
+            scope(exit) unlock();
+            for (;;)
+            {
+                if (toc[offset].empty)
+                {
+                    Data* target = blockAt(*nextAlloc);
+                    // Can't split a chunk if < itemSize + Data.sizeof
+                    while (target != dataEnd && (target.occupied || target.fullSize < itemSize + Data.sizeof))
+                    {
+                        target = next(target);
+                    }
+                    if (target == dataEnd) return false; // TODO: GC blocks
+                    toc[offset].hash32 = h;
+                    toc[offset].index = indexOfBlock(target);
+                    uint blockSize = target.fullSize;
+                    target.occupied = 1;
+                    target.keySize = cast(uint)key.length;
+                    target.valueSize = cast(uint)value.length;
+                    target.key[] = cast(ubyte[])key[];
+                    target.value[] = cast(ubyte[])value[];
+                    auto tail = next(target);
+                    tail.occupied = 0;
+                    tail.keySize = cast(uint)(blockSize - itemSize - Data.sizeof);
+                    tail.valueSize = 0;
+                    *nextAlloc = indexOfBlock(tail);
+                    return true;
+                }
+                offset = (offset + 1) & (toc.length-1);
+            }
+        }
+
+    private:
+
+        Data* blockAt(uint index)
+        {
+            return cast(Data*)(cast(ubyte*)data + index);
+        }
+
+        uint indexOfBlock(Data* block)
+        {
+            return cast(uint)(cast(ubyte*)block - cast(ubyte*)data);
+        }
+
+        static Data* next(Data* block)
+        {
+            return cast(Data*)(cast(ubyte*)block + block.fullSize);
+        }
+
+        void printToc()
+        {
+            printf("--- DCache TOC ---\n");
+            foreach(entry; toc)
+            {
+                if (!entry.empty)
+                    printf("hash: %x index: %d\n", entry.hash32, entry.index);
+            }
+            printf("------------------\n");
+        }
+
+        void printBlocks()
+        {
+            Data* p = data;
+            printf("--- DCache Blocks ---\n");
+            while(p != dataEnd)
+            {
+                printf("offset = %d occupied = %d keySize = %d valueSize = %d full = %d\n",
+                    indexOfBlock(p), p.occupied, p.keySize, p.valueSize, p.fullSize);
+                p = next(p);
+            }
+            printf("---------------------\n");
+        }
+
+        static void[] mapRange(size_t size, int fd)
+        {
+            void * p = mmap(null, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+            if (p == MAP_FAILED)
+            {
+                perror("Failed to mmap");
+                abort();
+            }
+            return p[0..size];
+        }
+
+        static uint first32bits(const(ubyte)[] key)
+        {
+            return *cast(uint*)key.ptr;
```

Comment:
Please avoid non-portable unaligned reads, just use `p[0] | p[1] << 8 | p[2] << 16 | p[3] << 24`, it will be optimized to a 32-bit read when the architecture has hardware support for unaligned reads.